### PR TITLE
fix: robust patch fetching to avoid a race condition

### PR DESCRIPTION
**Pull request by adrukh**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

Stream-based implementation of patch fetching used to have a race condition where the promise would `resolve` as soon as the patch content was downloaded, while letting the file write stream hanging to complete whenever it felt like. This resulted in sporadic patch failures where patch files were read before data was fully written, resulting in an attempt to apply an empty patch.

This is now fixed by making patch fetching logic more robust, summoning the power of TypeScript!

### CODE REVIEW

File: src/lib/protect/fetch-patch.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import * as fs from 'fs';
+import * as analytics from '../analytics';
+import * as debugModule from 'debug';
+import * as proxyFromEnv from 'proxy-from-env';
+import * as ProxyAgent from 'proxy-agent';
+
+const debug = debugModule('snyk:fetch-patch');
+const getProxyForUrl = proxyFromEnv.getProxyForUrl;
+
+async function getPatchFile(url, filename) {
```
Comment (michael-go) [inline src/lib/protect/fetch-patch.ts]: can add type annotations

File: src/lib/protect/fetch-patch.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  let options;
+  const proxyUri = getProxyForUrl(url);
+  if (proxyUri) {
+    debug('Using proxy: ', proxyUri);
+    options = { agent: new ProxyAgent(proxyUri) };
+  }
+
+  let res;
```
Comment (michael-go) [inline src/lib/protect/fetch-patch.ts]: if you install `@types/needle` you can then do 
```
let res: NeedleResponse;
```

File: src/lib/protect/fetch-patch.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  let patchData;
```
Comment (michael-go) [inline src/lib/protect/fetch-patch.ts]:
```typescript
let patchData: string;
```

File: src/lib/protect/fetch-patch.ts (hunk 1 above)
Comment (michael-go) [inline src/lib/protect/fetch-patch.ts]: would rename `filename` to `outFilename` or somethin'

and b.t.w. now that we use jsDiff - do we need to actually write this to file, or this function can just return the contents instead?

File: src/lib/protect/fetch-patch.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  try {
+    res = await needle('get', url, options || {});
```
Comment (michael-go) [inline src/lib/protect/fetch-patch.ts]: will `needle` throw if it's not 200?

File: src/lib/protect/fetch-patch.ts (hunk 1 above)
Comment (adrukh) [reply src/lib/protect/fetch-patch.ts]: I'd rather not make a functional change right now... The serial patching process is managed by a reducing promises chain, needs a refactor of its own :)

File: src/lib/protect/fetch-patch.ts (hunk 4 above)
Comment (adrukh) [reply src/lib/protect/fetch-patch.ts]: OMG it doesn't

File: src/lib/protect/fetch-patch.ts (hunk 2 above)
Comment (adrukh) [reply src/lib/protect/fetch-patch.ts]: Couldn't make it work for some reason...

Comment (michael-go) [reply src/lib/protect/fetch-patch.ts]: sorry, this:
```typescript
let res: needle.NeedleResponse;
```

Comment (adrukh) [reply src/lib/protect/fetch-patch.ts]: Done!

**Pull request merged**